sycl/test/fpga_tests/fpga_pipes.cpp

```diff
@@ -0,0 +1,207 @@
+// RUN: %clang -std=c++11 -fsycl %s -o %t.out -lstdc++ -lOpenCL -lsycl
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+//==------------- fpga_pipes.cpp - SYCL FPGA pipes test --------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include <CL/sycl.hpp>
+#include <iostream>
+
+// For pipes created with namespaces set
+namespace some {
+  class pipe;
+}
+
+using namespace cl::sycl;
+
+const unsigned dataSize = 1024;
+
+int main() {
+  int data[dataSize];
+  for (size_t i = 0; i != dataSize; ++i)
+    data[i] = 0;
+
+  {
+    // Test for non-blocking pipes
+    queue Queue;
+    using Pipe = pipe<class some_pipe, int, 1>;
+
+    Queue.submit([&](handler& cgh) {
+      cgh.parallel_for<class foo_nb>(range<1> { dataSize }, [=](id<1> idx) {
+        bool SuccessCode = false;
+        while (!SuccessCode)
+          Pipe::write(42, SuccessCode);
+      });
+    });
+
+    buffer<int, 1> writeBuf (data, range<1>(dataSize));
+    Queue.submit([&](handler& cgh) {
+      auto write_acc = writeBuf.get_access<access::mode::write>(cgh);
+
+      cgh.parallel_for<class goo_nb>(range<1> { dataSize }, [=](id<1> idx) {
+        bool SuccessCode = false;
+        while (!SuccessCode)
+          write_acc[idx] = Pipe::read(SuccessCode);
+      });
+    });
```

Comment:
Is there any user-friendly  documentation on how this Pipe extension works?
One of the things that confuses me is that, in the test above, there is nothing preventing a SYCL runtime to execute `goo_nb` before `foo_nb`, since the intersection of the requirements is empty and SYCL queues are out-of-order.

Is that relevant in your case? I figure for an FPGA it doesn't matter, but there is nothing in this extension that suggest it shouldn't work on anything that is not an FPGA but supports pipes.

Reply:
> Is there any user-friendly documentation on how this Pipe extension works?

It's going to be published. Don't really know when.
There is SYCL 2.2 spec (which is deprecated, but yet can provide some information). Also this implementation doesn't follows it.

Reply:
> I figure for an FPGA it doesn't matter, but there is nothing in this extension that suggest it shouldn't work on anything that is not an FPGA but supports pipes.

Yeah, I shall put some code in device information quarrying.

Reply:
We'll publish the extension document soon in this repo

Reply:
The problem is to come with a terse API and at the same time not giving up any performance on the table...
Perhaps a more explicit naming in the case of (non-)blocking? Adding a tag to pick another overload? Adding a template parameter?...
I understand that relying on an `accessor` to encode the blocking behavior like in SYCL 2.2 might be a too heavy syntax....

Reply:
The pipes extension at SYCL level is described in the document, currently as a pending PR, at:
https://github.com/intel/llvm/pull/635

Reply:
Ah thank you for the cross-reference. It prevents me from looking for it...